frame/transaction-payment/src/payment.rs

```diff
@@ -0,0 +1,124 @@
+///! Traits and default implementation for paying transaction fees.
+use crate::{NegativeImbalanceOf, Trait};
+use codec::FullCodec;
+use frame_support::{
+	traits::{Currency, ExistenceRequirement, Get, Imbalance, OnUnbalanced, WithdrawReason},
+	unsigned::TransactionValidityError,
+};
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, DispatchInfoOf, MaybeSerializeDeserialize, PostDispatchInfoOf, Saturating, Zero},
+	transaction_validity::InvalidTransaction,
+};
+use sp_std::{fmt::Debug, marker::PhantomData};
+
+/// Handle withdrawing, refunding and depositing of transaction fees.
+pub trait OnChargeTransaction<T: Trait> {
+	/// The underlying integer type in which fees are calculated.
+	type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default;
+	type LiquidityInfo: Default;
+
+	/// Before the transaction is executed the payment of the transaction fees
+	/// need to be secured.
+	///
+	/// Note: The `fee` already includes the `tip`.
+	fn withdraw_fee(
```

Comment:
I kind alike naming the two trait methods 

`fn pre_dispatch_withdraw()` and `fn post_dispatch_correction` but totally up to you.

Reply:
I really like the names. But I'm undecided. I'm not sure whether `pre_dispatch` and `post_dispatch` is that important for the Trait to put it into their function names. In the end, the Trait doesn't provide hooks to be called pre and post dispatch, but rather handle the withdraw and deposit of fees.